### DISCUSSION TITLE
Questions about training logs

### DISCUSSION THREAD

**Discussion by xxw11** [hpcaitech/ColossalAI#254] [Community | Q&A] [answered]
Hi, I want to know whether it is possible to get training information of each step including time cost, loss, learning rate?
Can I print the information or convey it to the tensorboard by adding a hook?

**Comment by FrankLeeeee** [comment]
Yes, this can be done regardless of whether you are using trainer or engine.

First of all, for general logging, you can use our distributed logger. The API documentation is [here](https://colossalai.readthedocs.io/en/latest/colossalai/colossalai.logging.html). The example can be found below.

```python
from colossalai.logging import get_dist_logger
from colossalai.context import ParallelMode

logger = get_dist_logger()

# this will create a log file for each rank
logger.log_to_file('./)

# log on all ranks
logger.info('hello world')

# log on only rank 0 in the global process group
logger.info('hello world', ranks=[0])

# log on only rank 0 in the data-parallel process group
logger.info('hello world', ranks=[0], parallel_mode=ParallelMode.DATA)


```

**Comment by FrankLeeeee** [comment, marked as answer]
If you are using a trainer, you can add these hooks

```python
from colossalai.trainer import hooks

hook_list = [
        hooks.LossHook(),
        hooks.LRSchedulerHook(lr_scheduler=lr_scheduler, by_epoch=True),
        hooks.LogMetricByEpochHook(logger),
        hooks.ThroughputHook(),
        hooks.LogMetricByStepHook(),
        hooks.TensorboardHook(log_dir='./tb_logs', ranks=[0]),
    ]
```

**Reply by FrankLeeeee** [reply]
If you want to save the logs produced by these hooks into a file, call `logger.log_to_file`. The `TensorboardHook` hook here will only store the stats on rank 0. This is generally good enough. The hooks documentation can be found [here](https://colossalai.readthedocs.io/en/latest/colossalai/colossalai.trainer.hooks.html).

**Reply by xxw11** [reply]
Hi, a loss hangging problem is occured
![image](https://user-images.githubusercontent.com/55141908/155644275-97243918-5020-4af0-8354-5e2063618360.png)

**Reply by ver217** [reply]
Hi, could you tell me the technique you use? Tensor parallel and pipeline parallel? Or ZeRO?

**Reply by xxw11** [reply]
Tensor parallel 2 and ZeRO2

**Reply by ver217** [reply]
We are trying to reproduce your issue. By the way, we are refactoring ZeRO now, and your issue may be solved when we publish a new version.

**Reply by FrankLeeeee** [reply]
We are currently not sure about the root cause as well as this is associated with your algorithm consideration. I will see if we can reproduce this loss hanging problem and it can take some time. By the way, if you can give us your configuration for zero, it will help us for reproduction.

**Reply by xxw11** [reply]
[gpt2_2d.txt](https://github.com/hpcaitech/ColossalAI/files/8138178/gpt2_2d.txt)

hi ,this is my config file

**Reply by xxw11** [reply]
emmm 为了保证达到小loss我应该去掉zero2重新训练  对吗

**Reply by FrankLeeeee** [reply]
hmmm you can try this only tensor parallel. We are not sure if setting attention mask to all 1 will lead to any performance issue.

**Reply by FrankLeeeee** [reply]
But setting attention mask to all 1s is against my common sense about NLP.

**Reply by xxw11** [reply]
“But setting attention mask to all 1s is against my common sense about NLP.”我或许是这样出问题了，我应该如何设置 attention mask呢

**Reply by xxw11** [reply]
我是有padding的地方设置为0 无padding的地方设置为 1 我看它似乎没有padding就全1了

**Reply by FrankLeeeee** [reply]
> 我是有padding的地方设置为0 无padding的地方设置为 1 我看它似乎没有padding就全1了

好的，我觉得这样的话应该没问题。可以先试一下不加zero吗？我们这边也先测测看。

**Reply by xxw11** [reply]
![image](https://user-images.githubusercontent.com/55141908/155651586-952fae40-f93b-47e0-96ee-e2a4229d9b3a.png)
不加zero后  确实没有这个loss hanging的问题  （但这样我不能在限定时间内跑完了555

**Reply by FrankLeeeee** [reply]
zero配置我看dynamic loss scale是False，如果是True，还会有这个问题吗？

**Reply by xxw11** [reply]
我已经在尝试了，但是loss目前一直为nan ，这也是我之前将它关闭的主要原因，应该很快会降下来？

**Reply by xxw11** [reply]
我的初始loss比较大（1400+ ）不知道会不会有影响  目前先放着跑了 不论如何 十分感谢你们的回复！

**Reply by xxw11** [reply]
![image](https://user-images.githubusercontent.com/55141908/155665787-9f0f5966-b7e8-4d8c-80df-42a2fd4a1c1d.png)
loss仍然是nan，这是正常的吗

**Reply by FrankLeeeee** [reply]
你有用fp16吗？

**Reply by xxw11** [reply]
关于fp16的注释掉了，zero2和fp16在colossal ai中似乎没法同时使用
![image](https://user-images.githubusercontent.com/55141908/155666545-ee0ff45a-e484-46b8-bb57-854c2c217185.png)

**Reply by FrankLeeeee** [reply]
这个是训练多久变的nan呢？

**Reply by xxw11** [reply]
在第一个iteration之后
![image](https://user-images.githubusercontent.com/55141908/155667083-2f91ac93-cdbe-4d50-9800-b3095946a82f.png)

**Reply by FrankLeeeee** [reply]
好的，我试试看看能不能复现这个问题。

**Reply by ver217** [reply]
你的vocab size是怎么样的呢，tokenizer是用的什么呢，以及你的loss是GPTLMLoss嘛？

**Reply by xxw11** [reply]
vocab_size 是53228
tokenizer用的是megatron的EncDecTokenizer（好像是封装了wordpiece的一个tokenizer）
我用的是example里面的train_gpt.py  我没有指定loss_fn 所以用的是那里面的GPTLMLoss()

**Reply by xxw11** [reply]
希望这些有助于colossal ai de掉这些bug
现在重跑的话，我个人的任务在单个机器上已经没有办法按时完成了。
不过实际我可用的机器有两台，各有4张显卡，没有共享的文件系统，但能进行ssh通信。
请问colossal ai 可以支持这样的分布式训练吗

**Reply by ver217** [reply]
支持的，数据集在每个节点分别存一份就行

**Reply by xxw11** [reply]
我比较困惑的是整体的训练速度会不会受限于ssh的通信速度，如果我使用3d tensor并行和fp16的话

**Reply by ver217** [reply]
对于 3D tensor 并行来说，通信开销会比较显著，如果通信带宽较低是会受限制的。这种情况有限考虑 pipeline 并行或者 1D tensor 并行。

**Reply by xxw11** [reply]
十分感谢你的回答，如果通信带宽足够，我想短时间训练大量数据（batch_size尽可能大）就理论而言  3D tensor并行是最佳的选择吗（

**Reply by xxw11** [reply]
如果可以的话，可以给一个使用colossal进行分布式训练的例子吗 ？十分感谢

**Reply by ver217** [reply]
我们建议在能使用流水线并行时（如一层能够放在一张显卡上），优先使用流水线并行，因为流水线并行的通信量通常远小于张量并行。这是使用流水线并行训练ViT的教程：
https://www.colossalai.org/tutorials/advanced_tutorials/train_vit_using_pipeline_parallelism
这是使用流水线并行+张量并行训练GPT的教程：
https://www.colossalai.org/tutorials/advanced_tutorials/train_gpt_using_hybrid_parallelism

**Reply by feifeibear** [reply]
@ver217 wwx11的截图显示它的optimizer type=SGD，1. 如果没用ADAM应该也会影响它的收敛速度吧。2. 不用ADAM似乎也没必要用zero吧？zero主要是对Optimizer states减少显存的。

**Reply by xxw11** [reply]
Hi,或许我能尝试解答部分问题
2.zero2 zero3 不只是优化Optimizer states减少显存，在我的实验中，即使是SGD仍然可以显著提高batch_size（8->12）,
1.应该会影响，因为SGD对学习率十分敏感，但不至于loss无法小于10.883 我后来关了zero之后，loss很快小于10.883

**Reply by ver217** [reply]
Hi, 我不知道你是训练什么模型，一般来讲训练GPT还是用Adam比较多

**Reply by xxw11** [reply]
Hi,我在训练中文的GPT2模型，事实上我在参加一个比赛，比赛评判标准是在一定范围内（loss<7 一般认为比较好达到），然后尽可能快的训练1B tokens，所以我才会出此下策

**Reply by feifeibear** [reply]
@xxw11 您的训练过程的一些参数配置似乎和常见的example并不相同，但是可以肯定的是采用zero应该和未采用zero的结果保持一致，这有可能是colossal-ai的bug导致的。
您可以把您的目前的问题代码放在github上？如果涉及私有数据可以换成公开的数据集，可以让colossal-ai开发人员帮助一起调试？

**Reply by xxw11** [reply]
好的，我十分愿意分享我的代码来使colossal-ai更加完善！
事实上我还是有一些十分困惑的地方，
1.多维tensor并行会使通信开销非常显著，那么它的优势是在于单机多卡（至少8张显卡）节省显存的，我这样的理解对吗？
2.多种并行策略给我带来了很多种选择也带来了疑惑，按照 @ver217 的说法，流水线并行主要用于节省通信成本，那么
我可以单台主机内4张卡2d tensor并行，主机间流水线并行
或者我单台主机4张卡2d tensor并行，主机间简单做数据并行
这两种方法似乎都可行，那种方法能够达到最快的效率呢？（pipeline并行和1d tensor并行似乎并不能体现colossal-ai的特点和优势，我的理解是colossal实现的多维tensor并行 是colossal十分占优的点）

**Reply by ver217** [reply]
1. 你理解的没错，一般tensor并行的process group都会选择在单机多卡
2. 能用数据并行的话，数据并行肯定是最快的。2d tensor 并行主要针对较大的模型，比如一层甚至没法放在一张卡，他切分方法导致比 1d tensor 并行更省内存。他们应对的其实是不太相同的场景。至于哪种并行方式速度快，还需要具体场景具体分析。就你举得例子来说，如果你的模型能用 pipeline 并行，我会优先使用 pipeline + data 并行。如果不行，我会考虑 tensor + data 并行。其实从实现上来讲，我们实现的pipeline 并行整体上也比 megatron 稍微快一点，而且通用性更强。

**Reply by binmakeswell** [reply]
感谢 @xxw11 对本项目的关注，欢迎加入Colossal-AI微信社群，与本项目保持长久的沟通联系  :-)
![Colossal-AI_WeChat](https://user-images.githubusercontent.com/61670638/156148072-6ce2241a-39c2-4178-86c4-3ba5615bdbfb.png)
